expmod/src/lib.rs

```diff
@@ -0,0 +1,329 @@
+extern crate ewasm_api;
+extern crate num_bigint;
+
+use num::{One, Zero};
+use num_bigint::{BigInt, Sign};
+use std::cmp;
+use std::ops::Shl;
+
+fn f(x: usize) -> u64 {
+    let cost = match x {
+        0...64 => x * x,
+        65...1024 => (x * x / 4) + 96 * x - 3072,
+        _ => (x * x / 16) + 480 * x - 199680,
+    };
+    cost as u64
+}
+
+const G_QUAD_DIVISOR: u64 = 20;
+
+fn calculate_cost(lm: usize, lb: usize, exp: &BigInt) -> u64 {
+    let le = exp.to_bytes_be().1.len();
+    let lep: u64 = match le {
+        0...32 => {
+            if exp.is_zero() {
+                0u64
+            } else {
+                // Number of bits == ⌊log2()⌋ + 1
+                exp.bits() as u64 - 1
+            }
+        }
+        _ => {
+            let low32bytes = exp.modpow(&BigInt::one(), &BigInt::one().shl(256));
+            let loglow32bytes = low32bytes.bits() as i64 - 1;
+            if low32bytes > BigInt::zero() {
+                8 * (le as u64 - 32) + loglow32bytes as u64
+            } else {
+                8 * (le as u64 - 32)
+            }
+        }
+    };
+
+    (f(cmp::max(lm, lb)) * cmp::max(lep, 1 as u64)) / G_QUAD_DIVISOR
+}
+
+fn expmod(base: &BigInt, exp: &BigInt, modulus: &BigInt) -> Vec<u8> {
+    let x = if modulus.is_zero() {
+        BigInt::zero()
+    } else {
+        base.modpow(exp, modulus)
+    };
+    let (_, mut data) = x.to_bytes_be();
+
+    /* padded on 32 bytes */
+    let mut padded = vec![0u8; (32 - (data.len() % 32)) % 32];
+    padded.append(&mut data);
+    padded
+}
+
+#[cfg(not(test))]
+#[no_mangle]
+pub extern "C" fn main() {
+    use num::cast::ToPrimitive;
+    use std::cmp::min;
+
+    const BASE_LENGTH_OFFSET: usize = 0;
+    const BASE_LENGTH_LENGTH: usize = 32;
+    const EXP_LENGTH_OFFSET: usize = BASE_LENGTH_LENGTH;
+    const EXP_LENGTH_LENGTH: usize = 32;
+    const MOD_LENGTH_OFFSET: usize = EXP_LENGTH_OFFSET + EXP_LENGTH_LENGTH;
+    const MOD_LENGTH_LENGTH: usize = 32;
+    const BASE_OFFSET: usize = MOD_LENGTH_OFFSET + MOD_LENGTH_LENGTH;
+
+    let input_size = ewasm_api::calldata_size();
+    // Geth will consider the input size to be 0 if it's less than
```

Comment:
Is there a test in the official suite for this? I'm not sure I've seen this in the EIP, though I haven't looked at it for a while.

Reply:
it's definitely not in the white paper, but it's in the codebase. Which EIP are you talking about?

Reply:
most of it depends on the output of `calldata_copy` and I suspect the BigInt library to be very inefficient so I don't want to pass those BigInts back and make things worse. Although it hasn't been submitted yet, I have tests for these in the geth PR.

Reply:
The one which introduced it: https://eips.ethereum.org/EIPS/eip-198

Reply:
Basically this code in geth could be thought as an optimisation. The first 96 bytes contains the data lengths, but calldata is expected to be infinitely zero padded. In case it is less than 96 bytes, no matter what length is stored in the first fields, the values will be zero anyway.

This assumes that `modexp(0, 0, 0) = 0`.

However, one part this could cause a problem is gas calculation.

Reply:
Here's the go piece: https://github.com/ethereum/go-ethereum/blob/master/core/vm/contracts.go#L173

Reply:
I'm aware of that piece of code since this is where this comment comes from. The gas calculation is correct once this line is in, since this is one of the things that are needed for a correct sync of Rinkeby (meaning, I compared all txs on rinkeby one by one with the original, and got the same result). Therefore, could you explain what your point is?

Reply:
just one extra point: this isn't an optimization because if it was, it would return `[0, 0, ... 0]` (32 times) whereas with this it will return an empty array. I skimmed through the EIP and it seems to me that this case isn't explicitely covered.